### PR TITLE
Lifetimed FFI provider

### PR DESCRIPTION
**Pull request by robertbastian**
This allows creating a `BufferProvider` from borrowed bytes.

### CODE REVIEW

File: ffi/diplomat/cpp/docs/source/provider_ffi.rst [hunk 1]
```diff
@@ -34,13 +34,15 @@
...
 
-    .. cpp:function:: static diplomat::result<ICU4XDataProvider, ICU4XError> create_from_byte_slice(const diplomat::span<uint8_t> blob)
+    .. cpp:function:: static diplomat::result<ICU4XDataProvider, ICU4XError> create_from_bytes(const diplomat::span<uint8_t> blob)
 
         Constructs a ``BlobDataProvider`` and returns it as an :cpp:class:`ICU4XDataProvider`.
 
         See the `Rust documentation for BlobDataProvider <https://unicode-org.github.io/icu4x-docs/doc/icu_provider_blob/struct.BlobDataProvider.html>`__ for more information.
 
 
+        Lifetimes: ``blob`` must live at least as long as the output.
```
Comment (robertbastian) [inline ffi/diplomat/cpp/docs/source/provider_ffi.rst:44]: Key achievement in here

File: ffi/diplomat/wasm/icu4x/lib/ICU4XDataProvider.js [hunk 2]
```diff
@@ -56,8 +56,6 @@ export class ICU4XDataProvider {
         throw new diplomatRuntime.FFIError(throw_value);
       }
     })();
-    buf_arg_blob.free();
```
Comment (robertbastian) [inline ffi/diplomat/wasm/icu4x/lib/ICU4XDataProvider.js:59]: And here

**Comment by robertbastian** [comment]
I realised this doesn't work for the same reason it doesn't work in Rust: if we introduce a lifetimed version of a data provider, there's nothing stopping a DataResponse to outlive it.

**Comment by robertbastian** [comment]
I realised this doesn't work for the same reason it doesn't work in Rust: if we introduce a lifetimed version of a data provider, there's nothing stopping a DataResponse to outlive it.

**Comment by robertbastian** [comment]
I realised this doesn't work for the same reason it doesn't work in Rust: if we introduce a lifetimed version of a data provider, there's nothing stopping a DataResponse to outlive it.

**Comment by sffc** [comment]
ICU4XDataProvider in FFI probablly shouldn't gain a lifetime parameter, but I think it should be fine to introduce an unsafe constructor from static bytes as originally proposed. Just document it very well and make it look scary.

**Comment by robertbastian** [comment]
C users will be used to memory management, but for JS/WASM this will be very unusual. It would be nice if we could exclude this constructor from WASM.

**Pull request closed without merging**